### PR TITLE
Add a menu entry to change the moonraker port when configuring the in…

### PR DESCRIPTION
**Pull request by cashew22**
This allow the user to select the moonraker port when configuring the integration

 
image

#15

### CODE REVIEW

**Comment by marcolivierarsenault** [comment]
#15

File: tests/test_config_flow.py [hunk 1]
```diff
@@ -52,3 +52,10 @@ async def test_tmp_failing_config_flow(hass):
 
     assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
     assert result["errors"] == {"base": "error"}
+
```
Comment (marcolivierarsenault) [inline tests/test_config_flow.py:55]: you should have a distinct test for this

File: tests/test_api.py [hunk 2]
```diff
@@ -9,7 +9,7 @@ async def test_connect_client():
     with patch("moonraker_api.MoonrakerClient"), patch(
         "moonraker_api.websockets.websocketclient.WebsocketClient.connect"
     ), patch("moonraker_api.websockets.websocketclient.WebsocketClient.disconnect"):
-        moonraker_api = MoonrakerApiClient("notaURL", None)
+        moonraker_api = MoonrakerApiClient("notaURL", None, port="1234")
```
Comment (marcolivierarsenault) [inline tests/test_api.py:12]: we probably want to test for a bit more.

- Good port
- Bad port
- No port
- Not an int

File: custom_components/moonraker/config_flow.py [hunk 3]
```diff
@@ -26,17 +26,22 @@ async def async_step_user(self, user_input=None):
 
         if user_input is not None:
             valid = await self._test_connection(user_input[CONF_URL])
-            if valid:
-                return self.async_create_entry(
-                    title=DOMAIN, data=user_input
-                )  # changer DOMAIN pour name
-            self._errors["base"] = "error"
+            if not valid:
```
Comment (marcolivierarsenault) [inline custom_components/moonraker/config_flow.py:29]:
```suggestion
            if not await self._test_connection(user_input[CONF_URL]):
```

File: custom_components/moonraker/config_flow.py [hunk 4]
```diff
@@ -26,17 +26,22 @@ async def async_step_user(self, user_input=None):
 
         if user_input is not None:
             valid = await self._test_connection(user_input[CONF_URL])
```
Comment (marcolivierarsenault) [inline custom_components/moonraker/config_flow.py:28]:
```suggestion
```

File: custom_components/moonraker/config_flow.py [hunk 5]
```diff
@@ -26,17 +26,22 @@ async def async_step_user(self, user_input=None):
...
-                )  # changer DOMAIN pour name
-            self._errors["base"] = "error"
+            if not valid:
+                self._errors["base"] = "error"
+                return await self._show_config_form(user_input)
 
-            return await self._show_config_form(user_input)
+            valid = await self._test_port(user_input[CONF_PORT])
+            if not valid:
+                self._errors[CONF_PORT] = "port_error"
```
Comment (marcolivierarsenault) [inline custom_components/moonraker/config_flow.py:35]: Feels weird that this one use a const and not `self._errors["base"] = "error"`

Comment (cashew22) [reply custom_components/moonraker/config_flow.py:35]: base mean that the error is on top of the config windows and if it's equal the name of a fields the error will display at that fields in the UI.

I think it's the other one that should use CONF_URL in fact, i'll add this

File: tests/test_api.py (hunk 2 above)
Comment (cashew22) [reply tests/test_api.py:12]: I added the check in test_config_flow, i'm not sure to understand what test_api is actually testing...

File: tests/test_config_flow.py [hunk 6]
```diff
@@ -51,4 +51,32 @@ async def test_tmp_failing_config_flow(hass):
...
+    assert result["errors"] == {CONF_URL: "error"}
+
+
+async def test_server_port_config_flow(hass):
+    """Test some port config flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    # Not an int
```
Comment (marcolivierarsenault) [inline tests/test_config_flow.py:63]: split in separate test

File: tests/test_config_flow.py [hunk 7]
```diff
@@ -51,4 +51,32 @@ async def test_tmp_failing_config_flow(hass):
...
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input={CONF_URL: "1.2.3.4", CONF_PORT: "1234wdw"}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["errors"] == {CONF_PORT: "port_error"}
+
+    # Too small
```
Comment (marcolivierarsenault) [inline tests/test_config_flow.py:70]: split in separate test

File: tests/test_config_flow.py [hunk 8]
```diff
@@ -51,4 +51,32 @@ async def test_tmp_failing_config_flow(hass):
...
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input={CONF_URL: "1.2.3.4", CONF_PORT: "32"}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["errors"] == {CONF_PORT: "port_error"}
+
+    # Too big
```
Comment (marcolivierarsenault) [inline tests/test_config_flow.py:77]: split in separate test

File: tests/test_config_flow.py [hunk 9]
```diff
@@ -51,4 +51,32 @@ async def test_tmp_failing_config_flow(hass):
         )
 
     assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
-    assert result["errors"] == {"base": "error"}
+    assert result["errors"] == {CONF_URL: "error"}
+
+
+async def test_server_port_config_flow(hass):
```
Comment (marcolivierarsenault) [inline tests/test_config_flow.py:57]: we should also have a test when it is empty

**Review by marcolivierarsenault — COMMENTED** [review]
I am concern if this works when port is left empty. 

I think we want it to work in that case?

File: custom_components/moonraker/__init__.py [hunk 10]
```diff
@@ -35,8 +35,11 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
         hass.data.setdefault(DOMAIN, {})
 
     url = entry.data.get(CONF_URL)
+    port = entry.data.get(CONF_PORT)
 
-    api = MoonrakerApiClient(url, async_get_clientsession(hass, verify_ssl=False))
+    api = MoonrakerApiClient(
```
Comment (marcolivierarsenault) [inline custom_components/moonraker/__init__.py:40]: Do we convert the port to a int, because I only see it as a string, and looks like we want a int

Comment (cashew22) [reply custom_components/moonraker/__init__.py:40]: It's converted by pythonmagie when calling MoonrakerApiClient()

**Review by marcolivierarsenault — APPROVED** [review]
🎩 

Great job
